### PR TITLE
Implement rate limite blob service

### PR DESCRIPTION
**Pull request by smnzhu**
closes https://github.com/dapperlabs/flow-internal/issues/1727

TODO:
- [x] Rebase and implement cli options once https://github.com/onflow/flow-go/pull/2841 is merged
- [x] Add metrics

### CODE REVIEW

**Review by koko1123 — APPROVED** [review]
Nice!

File: network/p2p/blob_service.go [hunk 1]
```diff
@@ -158,3 +171,33 @@ func (s *blobServiceSession) GetBlob(ctx context.Context, c cid.Cid) (blobs.Blob
...
+	}
+}
+
+func (r *rateLimitedBlockStore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	size, err := r.Blockstore.GetSize(ctx, c)
+	if err != nil {
+		return nil, err
+	}
+
+	err = r.limiter.WaitN(ctx, size)
```
Comment (peterargue) [inline network/p2p/blob_service.go:195]: `WaitN`'s godocs says
> WaitN blocks until lim permits n events to happen. It returns an error if n exceeds the Limiter's burst size, the Context is canceled, or the expected wait time exceeds the Context's Deadline. The burst limit is ignored if the rate limit is Inf.

I think this means that the burst limit must be >=max blob size. Can you add a warning in the godocs for `WithRateLimit`

**Review by peterargue — APPROVED** [review]
looks good. can you add a unit test to confirm get's are rate limited when configured

**Comment by smnzhu** [comment]
bors merge

**Comment by smnzhu** [comment]
bors merge

**Pull request merged**